pymc/gp/hsgp.py

```diff
@@ -0,0 +1,357 @@
+        L: Optional[Sequence[float]] = None,
+        c: Optional[float] = None,
+        drop_first: bool = False,
+        parameterization="noncentered",
+        *,
+        mean_func: Mean = Zero(),
+        cov_func: Covariance,
+    ):
+        arg_err_msg = (
+            "`m` and L, if provided, must be sequences with one element per active "
+            "dimension of the kernel or covariance function."
+        )
+
+        if not isinstance(m, Sequence):
+            raise ValueError(arg_err_msg)
+
+        if len(m) != cov_func.n_dims:
+            raise ValueError(arg_err_msg)
+        m = tuple(m)
+
+        if (L is None and c is None) or (L is not None and c is not None):
+            raise ValueError("Provide one of `c` or `L`")
+
+        if L is not None and (not isinstance(L, Sequence) or len(L) != cov_func.n_dims):
+            raise ValueError(arg_err_msg)
+
+        if L is None and c is not None and c < 1.2:
+            warnings.warn("For an adequate approximation `c >= 1.2` is recommended.")
+
+        parameterization = parameterization.lower().replace("-", "").replace("_", "")
+        if parameterization not in ["centered", "noncentered"]:
+            raise ValueError("`parameterization` must be either 'centered' or 'noncentered'.")
+        else:
+            self._parameterization = parameterization
+
+        self._drop_first = drop_first
+        self._m = m
+        self._m_star = int(np.prod(self._m))
+        self._L = L
+        self._c = c
+
+        super().__init__(mean_func=mean_func, cov_func=cov_func)
+
+    def __add__(self, other):
+        raise NotImplementedError("Additive HSGPs aren't supported.")
+
+    @property
+    def L(self):
+        if self._L is None:
+            raise RuntimeError("Boundaries `L` required but still unset.")
+        return self._L
+
+    @L.setter
+    def L(self, value):
+        self._L = pt.as_tensor_variable(value)
+
+    def prior_linearized(self, Xs: TensorVariable):
+        """Returns the linearized version of the HSGP, the Laplace eigenfunctions and the square
+        root of the power spectral density needed to create the GP.  This function allows the user
+        to bypass the GP interface and work directly with the basis and coefficients directly.
+        This format allows the user to create predictions using `pm.set_data` similarly to a
+        linear model.  It also enables computational speed ups in multi-GP models since they may
+        share the same basis.  The return values are the Laplace eigenfunctions `phi`, and the
+        square root of the power spectral density.
+
+        Correct results when using `prior_linearized` in tandem with `pm.set_data` and
+        `pm.MutableData` require two conditions.  First, one must specify `L` instead of `c` when
+        the GP is constructed.  If not, a RuntimeError is raised.  Second, the `Xs` needs to be
+        zero centered, so it's mean must be subtracted.  An example is given below.
```

Comment:
Info about the return values should be in its own section

Reply:
it's saying to get correct results you need to give `L` and zero center `Xs`, I think thats about inputs

Reply:
oh I marked the wrong lines - meant the two last lines of the previous paragraph..

Regardless, there should be a `Returns` section ;)

Reply:
Would you be willing to merge as is?  This is something that could be fixed in a later pull request.  A separate PR that went through all the methods and added a `Returns` section would be very manageable to write and review.